### PR TITLE
KodaDot - AssetsHub NFT indexer - Milestone 1 

### PR DESCRIPTION
**Pull request by yangwao**
#### Project Abstract

The AssetHub Indexer is a GraphQL service designed to streamline the interaction with Uniques, NFT non-fungible assets, and fungible assets on AssetHubs (Statemine/Statemint). This service is specifically tailored to serve as a robust data layer for NFT-oriented dApps, leveraging the simplicity of GraphQL and REST API.

In the current landscape, developers face challenges interacting with Uniques and Assets on Statemine due to the complexity and time-consuming nature of querying fungible and non-fungible assets on AssetHub. AssetHub Indexer aims to address these challenges by providing a user-friendly GraphQL interface, thus reducing the time and effort required to query these assets.

The key advantages of the AssetHub Indexer include
---
- its focus on ease of use,
- versatility for a broad range of use cases, and
- its commitment to bringing Assets Hub (Statemine & Statemint) utilization & documentation closer to developers.

Furthermore, the AssetHub Indexer is a completely decentralized, open-source solution that respects user privacy by not collecting user data.

**By reducing the time required to query assets and providing a more user-friendly interface, the AssetHub Indexer aims to foster the growth and development of the Web 3.0 ecosystem in Polkadot.**

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [x] **Level 2**:  Up to $30,000, 3 approvals

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`kodadot_assethub_nft_indexer_statemine_statemint.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).

### CODE REVIEW

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-nft-assets-hub-indexer-by-kodadot/3528/1

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the thorough application, @yangwao! I will share it with the rest of the committee. I just have two questions:

1. Since you mentioned that KodaDot is already using the indexer, could you expand on which parts of the project are already implemented and which are still work in progress? 
2. To simplify the evaluation, could you compress the list of deliverables? We don't need a separate deliverable for every event. For example, 3a and b, 4a to f and 6a to e could be one deliverable each.

**Comment by yangwao** [comment]
> * Since you mentioned that KodaDot is already using the indexer, could you expand on which parts of the project are already implemented and which are still work in progress?

You can have a realtime check here - https://github.com/kodadot/stick/milestone/1

> 2\. To simplify the evaluation, could you compress the list of deliverables? We don't need a separate deliverable for every event. For example, 3a and b, 4a to f and 6a to e could be one deliverable each.

From our experience, the best is to have micro deliverables that the w3f reviewer can quickly evaluate.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application, and sorry for the delay here. I'm generally happy to support the project, but could you reduce the price? The current FTE of 4, for example, seems too high to me, considering your current progress as well as the fact that so far, only three people contributed to the project.  Reducing the grant to a level 2 project would also make it easier to accept it ;-)

**Review by semuelle — COMMENTED** [review]
I agree with David. Twelve person months seems a lot for the scope.

**Comment by semuelle** [comment]
Are you planning to make any further changes to the application, @yangwao, or should we continue reviewing as is?

**Comment by yangwao** [comment]
> considering your current progress as well as the fact that so far, only three people contributed to the project

There will be more people working on this. 

> Reducing the grant to a level 2 project would also make it easier to accept it

Yes, for now, it makes sense to split.

> Are you planning to make any further changes to the application

Will be updating this week.

**Comment by keeganquigley** [comment]
Thanks @yangwao feel free to ping us when we should take another look at it.

**Comment by semuelle** [comment]
> Will be updating this week.

any updates, @yangwao?

**Comment by yangwao** [comment]
Sure. I've been OOO, will look at the upcoming days

**Comment by yangwao** [comment]
I have read and hereby sign the Contributor License Agreement.

**Comment by yangwao** [comment]
Hey, @semuelle @keeganquigley @Noc2, I've split our proposal into Milestones. 
Now, we will ask only for Milestone 1, which is already done live in production. 
It's used by [KodaDot](https://github.com/kodadot/nft-gallery), [PlutoWallet](https://github.com/RostislavLitovkin/PlutoWallet) and soon by others.

Ref
---
- https://github.com/kodadot/stick/milestone/1
- KodaDot using Stick in the production of Kusama Assets Hub and Polkadot Assets Hub
  - https://kodadot.xyz/ahk/explore/items
  - https://kodadot.xyz/ahp/explore/items

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+| 9.      | Metadata Support for Fungible Assets | Implement a handler to add metadata to a fungible asset event from the chain.                          |
+| 9c.   | Fungible Asset Allowlist Setup        |Setting up allows list-based indexing of fungible assets.                                                       |
+| 10a.  | Data Views Development                |Construction of data views for efficient access to indexed data.                                               |
+| 10b.  | Implementing Metadata Caching Layer |Develop and retry IPFS metadata if un-indexed by 7j.                                                               |
+| 11.    | Transfer of Collection Ownership    |Incorporate functionality to transfer collection ownership.                                                    |
+| 12a.  | Allocation of Initial Storage Bandwidth |Provide an initial bandwidth budget for IPFS item fetching (Payment is based on bandwidth scale).               |
+| 13b.  | Initial Storage Budget Allocation     |Allocate initial storage budget of $5,000 (Equivalent to 100GB storage at $5 per GB) for initial items minting, allowing for future data expansion.|
```
Comment (semuelle) [inline applications/kodadot_assethub_nft_indexer_statemine_statemint.md:184]: I'm afraid these [aren't covered by the grants program](https://grants.web3.foundation/docs/faq#what-activitiespositions-do-you-fund).

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+| 9.      | Fungible Assets Schema Creation      |Design and formulation of a GraphQL schema entity representing fungible assets.                                |
+| 9.      | Fungible Assets Force Creation        |Handlers will be developed to add system tokens like KSM/DOT into fungible assets.                             |
+| 9.      | Fungible Assets CREATE Event          |An event handler for indexing the creation of a fungible event from a chain, such as (RMRK/USDT) will be developed.|
+| 9.      | Metadata Support for Fungible Assets | Implement a handler to add metadata to a fungible asset event from the chain.                          |
+| 9c.   | Fungible Asset Allowlist Setup        |Setting up allows list-based indexing of fungible assets.                                                       |
+| 10a.  | Data Views Development                |Construction of data views for efficient access to indexed data.                                               |
+| 10b.  | Implementing Metadata Caching Layer |Develop and retry IPFS metadata if un-indexed by 7j.                                                               |
```
Comment (semuelle) [inline applications/kodadot_assethub_nft_indexer_statemine_statemint.md:181]: What does 'unindexed' mean here? Unpinned?

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+| 0b.   | Documentation                                       | Comprehensive inline code documentation and an explicit README file to guide the project setup and execution.|
+| 0c.   | Test Guidelines                                      | Testing will cover major functionality with unit tests and provide a guide for executing these tests.|
+| 0d.   | Docker Integration                                   | A Dockerfile will enable the project to run within a Docker container.                          |
+| 1.      | Backward Compatibility Maintenance             | Ensuring backward compatibility with current Uniques v1.                                                      |
+| 1a.   | Collection Schema Development              | Development of a GraphQL schema entity that represents the collection.                                        |
+| 1b.   | NFT Schema Development                       | Formulation and creation of a GraphQL schema entity representing Non-fungible tokens.                         |
+| 2.      | Forward Support Implementation             | Future support included for NFT pallets comprising soulbound NFTs.                                           |
```
Comment (semuelle) [inline applications/kodadot_assethub_nft_indexer_statemine_statemint.md:152]: What's exactly being done here? What does 'future support' mean?

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the update, @yangwao. Please see some questions below.

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md (hunk 1 above)
Comment (yangwao) [reply applications/kodadot_assethub_nft_indexer_statemine_statemint.md:184]: Sure, we can subtract that. I think this was counted in the first version of this revision of this proposal.

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md (hunk 2 above)
Comment (yangwao) [reply applications/kodadot_assethub_nft_indexer_statemine_statemint.md:181]: I think this means that we will store those data if particular IPFS hashes are not pinned by original author for better UX.

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md (hunk 3 above)
Comment (semuelle) [reply applications/kodadot_assethub_nft_indexer_statemine_statemint.md:152]: I'm afraid "future support" doesn't fit into the grants program either. You could apply for a maintenance grant to fund ongoing maintenance of the project once it's implemented, but not as part of a regular grant application.

**Comment by semuelle** [comment]
Please ping me here when the changes relevant to my requests have been made, @yangwao.

File: applications/kodadot_assethub_nft_indexer_statemine_statemint.md (hunk 3 above)
Comment (yangwao) [reply applications/kodadot_assethub_nft_indexer_statemine_statemint.md:152]: Sure, I can remove it.

Comment (vikiival) [reply applications/kodadot_assethub_nft_indexer_statemine_statemint.md:152]: @semuelle this was meant to support soulbond (non-transferable) nfts. 

> Why we used forward?

Because nfts can become sb after certain period of time, but we will change the wording.

**Comment by yangwao** [comment]
@semuelle please have check.

**Review by semuelle — COMMENTED** [review]
Hi @yangwao. Thanks for the updates. However, I am slightly confused as the issues listed in the M1 tracker have been closed for six weeks, and the commits referenced in them are largely from April and May. Since we don't do retroactive funding, I'd suggest you apply for M2 instead.

**Comment by vikiival** [comment]
> Since we don't do retroactive funding, I'd suggest you apply for M2 instead.

Hello @semuelle.
sorry for confusion this proposal is aimed to AssetHub NFT indexer with ArrowSquid, [which was released two days ago](https://twitter.com/subsquid/status/1701533774415286575).

>  I am slightly confused as the issues listed in the M1 tracker have been closed for six weeks

Yes that was a proof-of-work that we are able to deliver good quality software.

> largely from April and May

Yes, all this is implemented using FireSquid technology that would come to EOL in 3-4 months. 

**FireSquid vs. ArrowSquid aren't the same**

The ArrowSquid is using a batch processor, that requires totally different data handling as FireSquid. Due to the nature of the ArrowSquid we are able do the processing more effectively but need to rewrite the code from scratch.

**TL;DR**

Prior this proposal the M1 was aimed as a proof of concept to present the successful delivery by our team. We aim to use latest tech (ArrowSquid) that requires reimplemetation of the current indexer.

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the updates, @yangwao, and sorry for the late reply. Could you add the details of your previous comment regarding ArrowSquid into the application? Currently, it's not mentioned at all. Also, you removed the 5k USD allocation as deliverable, but the total costs remained the same. I would have assumed that it would be reduced correspondingly.

**Comment by semuelle** [comment]
pinging @yangwao & @vikiival

**Comment by vikiival** [comment]
> pinging @yangwao & @vikiival 

@semuelle is the proposal ready to be merged?

- https://github.com/yangwao/Grants-Program/pull/2

**Comment by keeganquigley** [comment]
Thanks @vikiival the proposal can't be merged until it reaches 3 approvals. However I see your changes haven't been merged yet by @yangwao, I think these changes need to be merged into the application first in order to move forward with it.

**Comment by yangwao** [comment]
Hello @semuelle @keeganquigley please have check

**Review by keeganquigley — CHANGES_REQUESTED** [review]
Thanks @vikiival @yangwao although looks like the application is still missing an [Overview](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#overview-1) section. Can you add this in?

Also, the numbering in the milestone tables is missing 6, therefore it doesn't make sense as is. Can you fix the numbering scheme and perhaps consolidate some of the deliverables so that there aren't so many? For example, 4a. - 4f. could probably be combined into one section, as well as 7a. - 7k. Also section 9 doesn't follow the same letter scheme that the rest of the deliverables have.

**Comment by yangwao** [comment]
> you removed the 5k USD allocation as deliverable, but the total costs remained the same. I would have assumed that it would be reduced correspondingly.

Hey thanks for your assumption @semuelle, [it was supposed to be removed as residues from the first version of the proposal](https://github.com/w3f/Grants-Program/pull/1864/commits/c8c41e6171e4fd63d3919891add82b7e81089bea), we've realized that implementation costs are **slightly higher in ArrowSquid**, than we've anticipated, thus we are taking max allowance on this stage to at least cover our implementation costs. 
Hope it's OK for you as we are a trusted delivery partner in the Polkadot ecosystem with 7 indexers [[1]](https://github.com/kodadot/rubick), [[2]](https://github.com/kodadot/click), [[3]](https://github.com/kodadot/snek), [[4]](https://github.com/kodadot/stick). The latest one will be used by Talisman and NovaWallet. So we see it as an ecosystem public good with utility.

> Can you add this in?

It's added now. 

> Can you fix the numbering scheme and perhaps consolidate some of the deliverables so that there aren't so many?

Improved now.

Let us know @keeganquigley when you review it.

**Review by keeganquigley — APPROVED** [review]
Thanks for the changes @yangwao I think it would be great to have an indexer for AssetHub. Also based on Kodadot's standing in the community I'm willing to go forward with it. I will mark the application as ready for review and ping the rest of the committee.

**Review by semuelle — APPROVED** [review]
Thanks for the updates, @yangwao. I am also willing to go ahead with it.

**Review by Noc2 — APPROVED** [review]
